[BUG] some parquet files written useing the GPU cannot be read on the GPU

**Describe the bug**
I ran into an issue recently where I was working with spark and rapids.  The output of a GPU write could not be read by the CPU code.  Sadly the data itself is sensitive so I cannot just hand it out.  But I am happy to work with individuals to try and reproduce the issue.

**Steps/Code to reproduce bug**
It looks like it has something to do the dictionary because when I use the parquer-mr tools to cat it I see the following exception.

```
$ java -cp ./target/parquet-tools-1.11.0.jar:./target/dependency/\* org.apache.parquet.tools.Main cat --debug ../../rapids-plugin-4-spark/gpu.snappy.parquet
....
org.apache.parquet.io.ParquetDecodingException: Can not read value at 273868 in block 0 in file file:/<>/gpu.snappy.parquet
	at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:251)
	at org.apache.parquet.hadoop.ParquetReader.read(ParquetReader.java:132)
	at org.apache.parquet.tools.command.CatCommand.execute(CatCommand.java:86)
	at org.apache.parquet.tools.Main.main(Main.java:223)
Caused by: org.apache.parquet.io.ParquetDecodingException: Can't read value in column [geo_location_event] required binary geo_location_event at value 273868 out of 275000, 133868 out of 135000 in currentPage. repetition level: 0, definition level: 0
	at org.apache.parquet.column.impl.ColumnReaderBase.readValue(ColumnReaderBase.java:523)
	at org.apache.parquet.column.impl.ColumnReaderImpl.readValue(ColumnReaderImpl.java:30)
	at org.apache.parquet.column.impl.ColumnReaderBase.writeCurrentValueToConverter(ColumnReaderBase.java:409)
	at org.apache.parquet.column.impl.ColumnReaderImpl.writeCurrentValueToConverter(ColumnReaderImpl.java:30)
	at org.apache.parquet.io.RecordReaderImplementation.read(RecordReaderImplementation.java:406)
	at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:226)
	... 3 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1372
	at org.apache.parquet.column.values.dictionary.PlainValuesDictionary$PlainBinaryDictionary.decodeToBinary(PlainValuesDictionary.java:122)
	at org.apache.parquet.column.values.dictionary.DictionaryValuesReader.readBytes(DictionaryValuesReader.java:83)
	at org.apache.parquet.column.impl.ColumnReaderBase$2$6.read(ColumnReaderBase.java:344)
	at org.apache.parquet.column.impl.ColumnReaderBase.readValue(ColumnReaderBase.java:503)
	... 8 more
org.apache.parquet.io.ParquetDecodingException: Can not read value at 273868 in block 0 in file file:<>/gpu.snappy.parquet

```

**Expected behavior**
I would expect the output of writing a parquet file could be read by spark and other similar tools.

**Environment overview (please complete the following information)**
This is using 0.13-SNAPSHOT (6351671f27c48753935dcee64f54e25b9885c52e) that I built myself.  I am running on Ubuntu.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@
 - PR #4054 Fixed JNI to deal with reduction API changes
 - PR #4052 Fix for round-robin when num_partitions divides nrows.
 - PR #4049 Fix `cudf::split` issue returning one less than expected column vectors
+- PR #4065 Parquet writer: fix for out-of-range dictionary indices
 
 
 # cuDF 0.12.0 (Date TBD)

--- a/cpp/src/io/parquet/page_dict.cu
+++ b/cpp/src/io/parquet/page_dict.cu
@@ -113,6 +113,7 @@ __device__ void GenerateDictionaryIndices(dict_state_s *s, uint32_t t)
     uint32_t *dict_data = s->col.dict_data + s->ck.start_row;
     const uint32_t *valid_map = s->col.valid_map_base;
     uint32_t num_dict_entries = 0;
+
     for (uint32_t i = 0; i < s->row_cnt; i += 1024) {
         uint32_t row = s->ck.start_row + i + t;
         uint32_t is_valid = (i + t < s->row_cnt && row < s->col.num_rows) ? (valid_map) ? (valid_map[row >> 5] >> (row & 0x1f)) & 1 : 1 : 0;
@@ -190,6 +191,7 @@ gpuBuildChunkDictionaries(EncColumnChunk *chunks, uint32_t *dev_scratch)
     while (s->row_cnt < s->ck.num_rows) {
         uint32_t frag_start_row = s->ck.start_row + s->row_cnt, num_dict_entries, frag_dict_size;
         FetchDictionaryFragment(s, s->col.dict_data, frag_start_row, t);
+        __syncthreads();
         num_dict_entries = s->frag.num_dict_vals;
         if (!t) {
             s->num_dict_entries = 0;
@@ -272,19 +274,28 @@ gpuBuildChunkDictionaries(EncColumnChunk *chunks, uint32_t *dev_scratch)
                 }
             }
             __syncthreads();
-            // At this point the order of dictionary is non-deterministic, so reorder the duplicate rows such that the lowest
-            // row number is the non-duplicate value
-            if (is_valid && is_dupe && next - 1 > row) {
-                atomicMin(&s->col.dict_index[next - 1], row);
-            }
-            __syncthreads();
-            if (is_valid && is_dupe && next - 1 > row) {
-                if (s->col.dict_index[next - 1] == row) {
-                    s->col.dict_index[next - 1] = row | (1u << 31);
-                    s->col.dict_index[row] = row;
+            // At this point, the dictionary order is non-deterministic, and we want insertion order
+            // Make sure that the non-duplicate entry corresponds to the lower row number
+            // (The entry in dict_data (next-1) used for duplicate elimination does not need
+            // to be the lowest row number)
+            bool reorder_check = (is_valid && is_dupe && next - 1 > row);
+            if (reorder_check) {
+                next = s->col.dict_index[next - 1];
+                while (next & (1u << 31)) {
+                    next = s->col.dict_index[next & 0x7fffffff];
                 }
             }
-            __syncthreads();
+            if (__syncthreads_or(reorder_check)) {
+                if (reorder_check) {
+                    atomicMin(&s->col.dict_index[next], row);
+                }
+                __syncthreads();
+                if (reorder_check && s->col.dict_index[next] == row) {
+                    s->col.dict_index[next] = row | (1u << 31);
+                    s->col.dict_index[row] = row;
+                }
+                __syncthreads();
+            }
         }
         __syncthreads();
         num_dict_entries = s->num_dict_entries;

